ttnn/cpp/ttnn/operations/moreh/moreh_adamw/moreh_adamw_pybind.cpp

```diff
@@ -0,0 +1,94 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "moreh_adamw_pybind.hpp"
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include "ttnn/cpp/pybind11/decorators.hpp"
+#include "ttnn/operations/moreh/moreh_adamw/moreh_adamw.hpp"
+#include "ttnn/types.hpp"
+
+namespace py = pybind11;
+
+namespace ttnn::operations::moreh::moreh_adamw {
+
+void bind_moreh_adamw_operation(py::module& module) {
+    bind_registered_operation(
+        module,
+        ttnn::moreh_adamw,
+        R"doc(moreh_adamw(param_in: ttnn.Tensor) -> ttnn.Tensor)doc",
+        // Add pybind overloads for the C++ APIs that should be exposed to python
+        // There should be no logic here, just a call to `self` with the correct arguments
+        // The overload with `queue_id` argument will be added automatically for primitive operations
+        // This specific function can be called from python as `ttnn.prim.adamw(param_in)` or
+        // `ttnn.prim.adamw(param_in, queue_id=queue_id)`
+        ttnn::pybind_overload_t{
+            [](const decltype(ttnn::moreh_adamw)& self,
+               const Tensor& param_in,
+               const Tensor& grad,
+               const Tensor& exp_avg_in,
+               const Tensor& exp_avg_sq_in,
+
+               float lr,
+               float beta1,
+               float beta2,
+               float eps,
+               float weight_decay,
+               uint32_t step,
+               bool amsgrad,
+
+               const std::optional<const Tensor> max_exp_avg_sq_in,
+               const std::optional<const Tensor> param_out,
+               const std::optional<const Tensor> exp_avg_out,
+               const std::optional<const Tensor> exp_avg_sq_out,
+               const std::optional<const Tensor> max_exp_avg_sq_out,
+               const MemoryConfig& mem_config,
```

Comment:
```suggestion
               const std::optional<ttnn::MemoryConfig>& memory_config,
```
should optional

Reply:
Updated.

Reply:
In this [cmt](https://github.com/tenstorrent/tt-metal/pull/12265#discussion_r1745345834), you ask Thanh to remove optional. So, what should we use?

Reply:
No optional:
- struct operation_attributes_t

optional:
- pybind


I think we should ask TT about `invoke`. It seems that binary uses optional, while unary does not.
- binary: https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_device_operation.hpp#L219
- unary: https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.hpp#L46

Reply:
As in pybind, it's optional, is there any benefit of using non-optional in `operation_attributes_t`? If not, I think it's upto us to choose. However, I think using optional at both places is more reasonable.

Reply:
I don’t think there’s a benefit to having non-optional in the struct operation_attributes_t.
However, I align with TT’s code style because there are many changes in the ttnn code, and if the code styles differ, it would require more time to understand the code and consider better styles each time.
This is not an agreed-upon detail but a personal opinion.

Reply:
I see it. My code is updated.

Reply:
I asked about invoke and received a response.

"We prefer optional<MemoryConfig> argument. When nullopt, we default to the memory config of the first argument.
There are still some operations which do not follow the guidance."